freedos/source/subst/subst/subst.1
.TH SUBST 1 "22 May 1995"
.\"
.\"    This is the documentation for the SWSUBST package, a SUBST/JOIN & more
.\"    Copyright (C) 1995  Steffen Kaiser
.\"
.\"    This program is free software; you can redistribute it and/or modify
.\"    it under the terms of the GNU General Public License as published by
.\"    the Free Software Foundation; either version 2 of the License, or
.\"    (at your option) any later version.
.\"
.\"    This program is distributed in the hope that it will be useful,
.\"    but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"    GNU General Public License for more details.
.\"
.\"    You should have received a copy of the GNU General Public License
.\"    along with this program; if not, write to the Free Software
.\"    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.\"  $Id: SUBST.1 1.2 1995/05/26 05:31:58 ska Exp ska $
.SH NAME
subst \- substitute a path by a drive letter.
.SH SYNOPSIS
\fBsubst\fP \fB/?\fP 

\fBsubst\fP \fBdr1\fP: \fBdr2\fP:\fBpath\fP

\fBsubst\fP \fBdr\fP: \fB/d\fP

\fBsubst\fP
.SH DESCRIPTION
\fIsubst\fP is designed to manipulate the CDS (\fBC\fPurrent
\fBD\fPirectory \fBS\fPtructure), so a path of one drive can be accessed
by a drive letter.

\fIsubst\fP is part of the \fIswsubst\fP package. It can be created, if
it does not already exist, by copying the executable \fBswsubst.exe\fP to
\fBsubst.exe\fP.
.SH OPTIONS
.IP \fB/?\fP 
Displays a little help screen.
.IP \fB/d\fP
Causes \fIsubst\fP to break off the relation between the drive letter \fBdr\fP and
the substituted path. If \fBdr\fP is JOIN'ed and not SUBST'ed, the relation
is been broken off, too.
.IP "\fBdr1: dr2:path\fP"
Causes \fIsubst\fP to substitute the \fBpath\fP on drive \fBdr2\fP with
the drive letter \fBdr1\fP. If \fBdr1\fP is already SUBST'ed or JOIN'ed,
this relation will be broken off. \fBpath\fP need not neither exist nor
fully qualified. It will be created as well as the full path relative to
the current working directory on drive \fBdr2\fP.
.IP "\fBwithout any parameter\fP"
This dumps a list onto the screen containing all SUBST'ed drives in
the form: X: => Y:\PATH; where Y:\PATH is the fully-qualified path and
by X: substituted path. If there is no SUBST'ed drive, no output
occurs.
.SH "SEE ALSO"
swsubst, join.
.SH "KNOWN BUGS"
.IP \(bu
Networked drives cannot be used.
.IP \(bu
Sometimes substitutions onto a floppy drive won't work for unknown
reason.
.SH CONTRIBUTERS
    Steffen Kaiser
.br
Ernst-Th\("almann-Stra\("se 2
.br
     D-09126 Iden
.br
 Deutschland - Germany


gnu/ChangeLog
Sun Feb  2 21:22:01 1992  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* Version 1.2.

	* nl.c: Support multiple files and "-" for stdin.
	(main): Check for read and write errors.
	(nl_file): New function.

Wed Jan 29 10:09:10 1992  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tr.c (main): -t option was called -b in getopt spec.
	(validate): Don't warn that set1 is longer than set2.

	* tr.c: Rename --sysv-string2-truncate to --truncate-string1.

Fri Jan 17 16:29:05 1992  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* nl.c: New program from bin-src.

	* nl.c (main): Use a struct linebuffer for line_buf.
	(process_file): Use readline instead of fgets, to preserve NULs.
	(check_section): Use memcmp instead of strncmp.
	(proc_text): Print line_buf with fwrite instead of printf.

	* nl.c (main): Usage message if too many args given.  Check
	for error in closing input file.  Lengths of section delimiter
	strings were 1 too large.  Take separator_str into account in
	length of print_no_line_fmt.
	(build_print_fmt): Allocate space for print_fmt, in case
	separator_str is long.
	(proc_text): A blank line is one that contains nothing, not
	even nonprinting characters.

Fri Jan 17 01:04:22 1992  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* All programs: Document `--' instead of `+' to introduce
	long-named options, in usage messages.

	* sum.c (bsd_sum_file): Renamed from sum_file.
	(sysv_sum_file): New function.
	(main): Recognize an option to select between the algorithms.

Sun Jan  5 17:41:18 1992  Jim Meyering (meyering at churchy.gnu.ai.mit.edu)

	* pr.c (close_file, print_page): Fixed bug that had the command
	yes |head |pr -t printing "yyyyyyyyyy".
	* (print_page): Fixed bug that had pr -3 -a printing two too few
	trailer lines per page.
	* (main): Added restriction that -a and -m are incompatible.
	Although the POSIX spec doesn't explicitly say they shouldn't
	be used together, it says -a modifies the -column option and
	that -column shouldn't be used with -m.

Thu Jan  2 15:23:59 1992  David J. MacKenzie  (djm at albert.gnu.ai.mit.edu)

	* nl.c: Include regex.h after, not before, sys/types.h.

Thu Jan  2 12:18:10 1992  Tom Lord  (lord at geech.gnu.ai.mit.edu)

	* sort.c (fillbuf) return bytes buffered instead of bytes read.

Fri Dec 27 22:53:36 1991  Jim Kingdon  (kingdon at geech.gnu.ai.mit.edu)

	* sort.c (LINEALLOC): New #define.
	(struct lines): New field ``limit''.
	(initlines): Set it from new arg ``limit''.
	(sort, mergefps, checkfp): Pass new arg to initlines().
	(findlines): Don't realloc past lines->limit.

Tue Dec 24 01:24:03 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tac.c, sort.c, csplit.c, system.h: Change POSIX ifdefs to
	HAVE_UNISTD_H and _POSIX_VERSION.

	* xwrite.c: Change POSIX ifdef to HAVE_UNISTD_H.

Sat 14 Dec 1991 11:46:42  Jim Meyering  (meyering at wombat)

	* tr.c: Fixed an inaccurate comment on posix_pedantic.

Thu 12 Dec 1991 21:15:20  Jim Meyering  (meyering at hal)

	* tr.c: Changed underscores to hyphens in long option name
	"sysv_string2_truncate".

Wed Dec 11 13:33:34 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tac.c (main): Set obscure_syntax to tell re_search to
	allocate memory for the group registers.

Fri Dec  6 18:26:27 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tac.c, sort.c, csplit.c [POSIX]:  Use sigaction instead of
	signal, which POSIX doesn't have.
	* sort.c: Replace inthandler and huphandler with sighandler.
	* csplit.c (main): Only handle signals if they were not being
	ignored.

	* tr.c: POSIX_ME_HARDER -> POSIXLY_CORRECT.

Wed Dec  4 00:47:47 1991  Jim Meyering  (meyering at wombat)

	* tr.c (unquote): Reformat code so it doesn't go beyond column 80.
	* tr.c (squeeze_filter): Comment a little on why it's better
	to step through the input by two.
	* tr.c (set_initialize): Write a comment describing the function.
	* tr.c: Eliminated the variable `portability_warnings' and replaced
	references to it by references to `!posix_pedantic'.  One of the
	uses of portability_warnings had been wrong.

Tue Dec  3 14:03:35 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tr.c: New program.

Sun Dec  1 15:07:35 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* linebuffer.[ch] (freebuffer): New function (used by cron).

Thu Oct 17 22:30:22 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* system.h, configure, Makefile.in: Don't include memory.h if
	STDC_HEADERS, removing need for MEMORY_H_MISSING.

Thu 17 Oct 1991 16:53:07  Jim Meyering  (meyering at wombat)

	* pr.c (print_page): REALLY fixed `extra newline at EOF' problem.
	Somehow, part of my patch didn't make it last time.

Sat Oct 12 12:04:47 1991  David J. MacKenzie  (djm at churchy.gnu.ai.mit.edu)

	* tail.c (pipe_lines, pipe_bytes): Initialize `first->next'.

	* cmp.c (cmp): Print EOF message to stderr, not stdout, for
	POSIX 1003.2.11.2.

	* sort.c (xfwrite): fwrite never returns < 0, so check if
	number written is number we asked to write.
	(fillbuf, main): fread never returns < 0, so check ferror instead.
	From Rainer Orth.

Tue Oct  8 18:07:08 1991  Jim Meyering  (meyering at churchy)

	* pr.c (print_page): Really fixed `extra newline at EOF' problem.
	* (store_columns): Fixed bug that caused `pr -b -2' to coredump
	on files of certain lengths.

Fri Oct  4 22:30:25 1991  Jim Meyering  (meyering at churchy)

	* pr.c (print_page): Fixed to not add single spurious newline
	at EOF when using -t.

Wed Oct  2 01:02:05 1991  David J. MacKenzie  (djm at apple-gunkies)

	* pr.c (print_page): Don't pad the page if -t given.

	* csplit.c (load_buffer), sort.c (mergefps): Use bcopy, not memcpy.

	* memchr.c (memchr): Fix typo.

Thu Sep 26 12:35:00 1991  David J. MacKenzie  (djm at churchy.gnu.ai.mit.edu)


freedos/source/subst/subst.c
/*
    SWSUBST: Alternate CDS manipulator for MS-DOS
    Copyright (C) 1995-97,2002 Steffen Kaiser

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 1, or (at your option)
    any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* $RCSfile: subst.c $
   $Locker: ska $	$Name:  $	$State: Exp $

	Subst/join.

*/
#pragma hdrfile "simple.csm"

#include <string.h>
#include <dos.h>

#include "swsubst.h"
#include "yerror.h"

#ifndef lint
static char rcsid[] = "$Id: subst.c 3.2 2002/11/28 06:20:27 ska Exp ska $";
#endif

static DPB FAR *lw2dpb(int dr)
/* locate the DPB for a drive;
   dr: 'A' == A:
   return NULL on error
*/
{	DPB FAR *dpb;

    dr -= 'A';
    if((dpb = firstDPB) != NULL) {
	    while(dr-- && FP_OFF(dpb) != 0xffff) 
	        if (_osmajor < 4) 
	            dpb = dpb->vers.dos3.next;
	        else
	            dpb = dpb->vers.dos4.next;
	    if(FP_OFF(dpb) == 0xffff) dpb = NULL;
	 }
	 return dpb;
}

void switchSubst(int dr, char *path, int bslashoff)
/* SUBST/JOIN physical path to drive dr;
   automatically breaks existing SUBST/JOIN
*/
{	int flags;
	int pathLen;
	int root;	/* SUBST/JOIN to root drive */
	int locDir;	/* path points to a local drive */
	CDS FAR *dir;
	char *p;

	/* flip '/' to '\\' */
	p = path;
	while((p = strchr(p, '/')) != NULL)
		*p++ = '\\';
	/* strip trailing '\\' */
	pathLen = strlen(path);
	if(path[pathLen - 1] == '\\' && path[pathLen - 2] != ':')
		path[--pathLen] = '\0';

	if(creatDir) /* creat path */
		if(!mkPhysDir(*path - 'A', path + 2))
	 		warning(E_mkDir, path);

	if(0 != (locDir = path[1] == ':' && path[2] == '\\')) {
		if(*path >= 'a' && *path <= 'z') *path += 'A' - 'a';
		else if(*path < 'A' || *path > 'Z') locDir = 0;
	}
	root = locDir && path[3] == NUL;	/* root directory */
	if(bslashoff && bslashoff > pathLen)
		bslashoff = 0;

	if(sizeof(dir->current_path) <= pathLen)
		error(E_pathLen, path);

	if((dir = cds(dr)) == NULL)
		fatal(E_cds, 'A' + dr);

   	if(join) {	/* force re-read of drive, so DOS detect JOIN */
   		CDS FAR *ddir;

		if(!locDir) error(E_joinLocal, dr + 'A', path);

        if((ddir = cds(*path - 'A')) == NULL)
            fatal(E_cds, *path);
        ddir->u.LOCAL.start_cluster = 0xffff;
	}

	/* update data area */
	_fmemcpy(dir->current_path, path, pathLen + 1);
	dir->backslash_offset = bslashoff? bslashoff: join? 2: root? pathLen - 1: pathLen;
	dir->u.LOCAL.start_cluster = 0xffff;
   	dir->dpb = lw2dpb(join? dr + 'A': *path);
   	flags = dir->flags;	/* keep old flags, to detect, if #Joined changed */
   	dir->flags &= ~(JOIN | SUBST);

	if(root && *path == dr + 'A') {
		/* SUBST/JOIN X: => X:  -> neither SUBST nor JOIN necessary, break it */
		if(!dir->dpb && !(dir->flags & NETWORK)) {
			/* no DPB => PHYSICAL this is possibly no vaild physical drive */
			dir->flags &= ~PHYSICAL;
		}
		else dir->flags |= PHYSICAL;	/* make it available, this is 
			necessary, if a former OFF setting must be reversed. */
	}
	else {
		/* perform SUBST/JOIN */
		dir->flags |= !locDir? PHYSICAL :join? JOIN | PHYSICAL: SUBST | PHYSICAL;
	}

	if((flags ^ dir->flags) & JOIN) { /* add|sub JOIN => upd #joined */
		if(nrJoined == NULL) 
			warning(E_nrJoin);
		else {
			if(flags & JOIN) --*nrJoined;
			else ++*nrJoined;
		}
	}
}
